Code/Mantid/Framework/Algorithms/src/ChangeTimeZero.cpp

```diff
@@ -0,0 +1,324 @@
+#include "MantidAlgorithms/ChangeTimeZero.h"
+#include "MantidAlgorithms/CloneWorkspace.h"
+#include "MantidAlgorithms/ChangePulsetime.h"
+#include "MantidAPI/IMDIterator.h"
+#include "MantidAPI/IEventWorkspace.h"
+#include "MantidDataObjects/EventList.h"
+#include "MantidKernel/System.h"
+#include "MantidKernel/TimeSeriesProperty.h"
+#include "MantidKernel/PropertyWithValue.h"
+#include "MantidKernel/ArrayProperty.h"
+#include "MantidKernel/DateAndTime.h"
+#include "MantidKernel/DateTimeValidator.h"
+
+#include <boost/lexical_cast.hpp>
+#include <boost/shared_ptr.hpp>
+
+namespace Mantid {
+namespace Algorithms {
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(ChangeTimeZero)
+
+using namespace Mantid::Kernel;
+using namespace Mantid::API;
+using std::size_t;
+
+//----------------------------------------------------------------------------------------------
+/** Constructor
+ */
+ChangeTimeZero::ChangeTimeZero()
+    : m_isRelativeTimeShift(false), m_isAbsoluteTimeShift(false),
+      m_dateTimeValidator(boost::make_shared<DateTimeValidator>()), m_defaultTimeShift(0.0), m_defaultAbsoluteTimeShift("") {}
+
+//----------------------------------------------------------------------------------------------
+/** Destructor
+ */
+ChangeTimeZero::~ChangeTimeZero() {}
+
+//----------------------------------------------------------------------------------------------
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void ChangeTimeZero::init() {
+  declareProperty(new WorkspaceProperty<MatrixWorkspace>("InputWorkspace", "",
+                                                         Direction::Input),
+                  "An input workspace.");
+  declareProperty<double>("RelativeTimeOffset", m_defaultTimeShift,
+                  "A relative time offset in seconds.");
+
+  declareProperty("AbsoluteTimeOffset", m_defaultAbsoluteTimeShift,
+                  "An absolute time offset as an ISO8601 string.");
+
+  declareProperty(new WorkspaceProperty<MatrixWorkspace>("OutputWorkspace", "",
+                                                         Direction::Output),
+                  "An output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void ChangeTimeZero::exec() {
+
```

Comment:
You must include progress reporting. This could be a long-running algorithm so we must see what's happening. We may also need the ability to cancel the algorithm. See Progress.h.